### PR TITLE
Add ability to cancel the export

### PR DESCRIPTION
**Pull request by albinekb**
- Currently only works for mp4, need to update all converters in scripts/converter since the others have multi step-conversions, which needs to be cancelled correctly
- Not designed yet, just added a simple button

Opening this now to start a discussion, is this an acceptable way to do it? Any ideas on how this could be done better? 🤔

### CODE REVIEW

**Comment by sindresorhus** [comment]
Some quick thoughts on how to do the multi-step process: https://gist.github.com/sindresorhus/b78ff84403b6dfd4a3098c3f9ed81437 I wish we had some utility method on `p-cancelable` to combine multiple cancelables.

File: app/src/main/share-service-context.js [hunk 1]
```diff
@@ -12,7 +12,13 @@ export default class ShareServiceContext {
 
     this.filePath = () => convert(Object.assign({}, exportOptions, {
       defaultFileName: this.defaultFileName
-    }));
+    })).catch(err => {
+      if (err.canceled) {
```
Comment (sindresorhus) [inline app/src/main/share-service-context.js]: I think this should be consistent and check the error name like done elsewhere here. Or maybe we should add a `canceled` property to `p-cancelable` errors? Could be nice for convenience.

File: app/src/main/convert.js [hunk 2]
```diff
@@ -37,6 +37,23 @@ export default async function (exportOptions) {
     }
   });
 
+  ipcMain.once('cancel-processing', () => {
+    converter.cancel();
+    app.kap.mainWindow.send('export-progress', {text: ''});
```
Comment (sindresorhus) [inline app/src/main/convert.js]: The above line is repeated 3 times in this file. Would be nice to dry it up.

File: app/src/main/convert.js [hunk 3]
```diff
@@ -37,6 +37,23 @@ export default async function (exportOptions) {
...
+  });
+
+  try {
+    await converter;
+  } catch (err) {
+    if (converter.canceled) {
+      app.kap.mainWindow.send('export-progress', {text: ''});
+      err.canceled = true;
+      throw err;
```
Comment (sindresorhus) [inline app/src/main/convert.js]: This line is moot as it could just continue to the below `throw`.

File: app/src/main/convert.js (hunk 2 above)
Comment (sindresorhus) [reply app/src/main/convert.js]: Actually, do we really need to call this here as it's already called on line 49?

File: app/src/renderer/js/main.js [hunk 4]
```diff
@@ -331,6 +332,10 @@ document.addEventListener('DOMContentLoaded', () => {
     this.classList.toggle('is-active');
   };
 
+  progressBarCancelBtn.onclick = function () {
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: Arrow-function, since `this` is not used.

Comment (sindresorhus) [reply app/src/renderer/js/main.js]: Actually, let's start using `.addEventListener()`. It's better.

Comment (albinekb) [reply app/src/renderer/js/main.js]: Will do 👌

File: app/src/main/share-service-context.js (hunk 1 above)
Comment (albinekb) [reply app/src/main/share-service-context.js]: I can make it check for the name, and yes, that'd be nice 👍

**Comment by albinekb** [comment]
Closing in favour of #335 🙏

**Pull request closed without merging**